pkg/activator/net/revision_backends.go

```diff
@@ -0,0 +1,416 @@
+	}
+
+	svcPort, ok := GetServicePort(rw.protocol, svc)
+	if !ok {
+		return false, "", fmt.Errorf("unable to find port in service %s/%s", svc.Namespace, svc.Name)
+	}
+
+	// Context used for our probe requests
+	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
+	defer cancel()
+
+	dest := net.JoinHostPort(svc.Spec.ClusterIP, strconv.Itoa(svcPort))
+	httpDest := url.URL{
+		Scheme: "http",
+		Host:   dest,
+	}
+	ok, err := prober.Do(ctx, rw.transport, httpDest.String(),
+		prober.WithHeader(network.ProbeHeaderName, queue.Name),
+		prober.ExpectsBody(queue.Name))
+
+	return ok, dest, err
+}
+
+func (rw *revisionWatcher) probePodIPs() (map[string]bool, error) {
+	// Context used for our probe requests
+	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
+	defer cancel()
+
+	var probeGroup errgroup.Group
+	healthStatesCh := make(chan destHealth, len(rw.dests))
+
+	for _, dest := range rw.dests {
+		// If the dest is already healthy then save this
+		if curHealthy, ok := rw.healthStates[dest]; ok && curHealthy {
+			healthStatesCh <- destHealth{dest, true}
+			continue
+		}
+
+		pDest := dest
+		probeGroup.Go(func() error {
+			httpDest := url.URL{
+				Scheme: "http",
+				Host:   pDest,
+			}
+			ok, err := prober.Do(ctx, rw.transport, httpDest.String(),
+				prober.WithHeader(network.ProbeHeaderName, queue.Name),
+				prober.ExpectsBody(queue.Name))
+			healthStatesCh <- destHealth{pDest, ok}
+			return err
+		})
+	}
+
+	if err := probeGroup.Wait(); err != nil {
+		return nil, err
+	}
+	close(healthStatesCh)
+
+	healthStates := make(map[string]bool, len(rw.dests))
+	for dh := range healthStatesCh {
+		healthStates[dh.dest] = dh.health
+	}
+
+	return healthStates, nil
+}
+
+// checkDests performs probing and potentially sends a dests update. It is
+// assumed this method is not called concurrently.
+func (rw *revisionWatcher) checkDests() {
+	if !rw.revisionReady() {
+		if rw.clusterIPHealthy {
+			// we have a healthy clusterIP but revision is not ready. We must have scaled down.
+			rw.clusterIPHealthy = false
+
+			// Send update that were now inactive
+			rw.updateCh <- &RevisionDestsUpdate{Rev: rw.rev}
+		}
+
+		// We know this revision cannot be healthy so short circuit
+		return
+	}
+
+	if rw.clusterIPHealthy {
+		// cluster IP is healthy and we havent scaled down, short circuit
+		return
+	}
+
+	// First check the clusterIP
+	svc, err := rw.getK8sPrivateService()
+	if err != nil {
+		rw.logger.Errorw("Failed to lookup private service for revision", zap.Error(err))
+		return
+	}
+
+	// If clusterIP is healthy send this update and we are done
+	if ok, dest, err := rw.probeClusterIP(svc); err != nil {
+		rw.logger.Errorw(fmt.Sprintf("Failed to probe clusterIP %s/%s", svc.Namespace, svc.Name), zap.Error(err))
+	} else if ok {
+		rw.clusterIPHealthy = true
+		rw.healthStates = make(map[string]bool)
```

Comment:
I think I'm a bit confused here. Why are we not adding `dest` to `rw.healthStates` here and mark it healthy?

Reply:
healthStates is only for tracking the health of podIPs, when a clusterIP is healthy (the case in this branch) we assume all addreses in the endpoints object are healthy.